### DISCUSSION TITLE
How is pipelining and cluster with multiple shards handled by the go-redis library

### DISCUSSION THREAD

**Discussion by bibicu123** [redis/go-redis#2597] [Q&A] [answered]
I am using go-redis v8.11.5 with an AWS MemoryDB Redis instance
I wonder how does the go-redis library deal with commands in the pipeline that are hashed to different shards?

Imagine a pipeline with 2 commands:
keys {aaa}*
keys {bbb}*

where aaa and bbb hash to different shards/nodes.
Is the go-redis library ClusterClient smart enough to send first command to one node and the second command to another node?

Is the new v9 better at handling this situation?

**Comment by marcind** [comment, marked as answer]
`ClusterClient` will group commands by the nodes that their [keys shard to](https://github.com/redis/go-redis/blob/v8/cluster.go#L1098) and [issue requests to each node in parallel](https://github.com/redis/go-redis/blob/v8/cluster.go#L1115-L1132). The v9 client does pretty much the same thing: https://github.com/redis/go-redis/blob/v9/cluster.go#L1203-L1238.

**Reply by bibicu123** [reply]
Thanks for replying.

We are debugging an issue with AWS MemoryDB for Redis : since we changed from one shard to 2 shards, we get inconsistent values for some keys. They are returned as zero for some processes and with the correct value for other processes.

The AWS MemoryDB exposes one cluster endpoint, so to construct a new cluster client we pass something like:
 
Addrs:   []string{"clustercfg.xxx.memorydb.eu-central-1.amazonaws.com:6379"}
in ClusterOptions

Will the client library auto detect the 2 separate nodes and shards with their corresponding slots or do we have to manually pass a ClusterSlots function describing the slot to node mapping ?

**Reply by axilis-marko** [reply]
@marcind thanks for your answer. @bibicu123 Have you figured out an answer to your follow up question?

**Reply by axilis-marko** [reply]
Just a random update here. I had some issues with AWS Elasticache and cluster slots. It turned out I was initiating another non-cluster client which was causing the issues.
